### PR TITLE
fetchYarnDeps: fix by deduplicating requests

### PR DESCRIPTION
**Pull request by feathecutie**
## Description of changes

This fixes the `fetchYarnDeps` function used for packaging [yarn](https://yarnpkg.com/)-based projects by ensuring that each individual dependency is only fetched once.

I ran into an issue while trying to package https://github.com/linkeddata/dokieli, which has a `yarn.lock` file with two entries resolving to the same `github.com` URL:

```
"green-turtle@git://github.com/bergos/green-turtle.git#master", "green-turtle@https://github.com/csarven/green-turtle#master":
  version "1.4.0"
  resolved "https://github.com/csarven/green-turtle#081ae1c39591e7502ac5f9c121253e18122d2a71"
```

Since `fetchYarnDeps` internally uses `nix-prefetch-git` in order to fetch GitHub dependencies, the same `nix-prefetch-git` command was run twice, resulting in the following error:

```log
[...]
downloading https://registry.yarnpkg.com/graceful-fs/-/graceful-fs-4.2.11.tgz
downloading https://registry.yarnpkg.com/graceful-fs/-/graceful-fs-4.2.11.tgz
downloading https://github.com/csarven/green-turtle
downloading https://github.com/csarven/green-turtle
downloading https://registry.yarnpkg.com/has-bigints/-/has-bigints-1.0.2.tgz
downloading https://registry.yarnpkg.com/has-bigints/-/has-bigints-1.0.2.tgz
downloading https://registry.yarnpkg.com/has-flag/-/has-flag-3.0.0.tgz
Error: Command failed: nix-prefetch-git --out csarven_green_turtle.tmp --url https://github.com/csarven/green-turtle --rev 081ae1c39591e7502ac5f9c121253e18122d2a71 --builder
fatal: cannot copy '/nix/store/p45c31p2qhwjy4vf60v2d2mw9bvvpb67-git-2.45.2/share/git-core/templates/hooks/commit-msg.sample' to '/nix/store/9nyg6qj21776n21jvrcww9q0bi22bwxh-offline/csarven_green_turtle.tmp/.git/hooks/commit-msg.sample': File exists

    at genericNodeError (node:internal/errors:984:15)
    at wrappedFn (node:internal/errors:538:14)
    at ChildProcess.exithandler (node:child_process:422:12)
    at ChildProcess.emit (node:events:519:28)
    at maybeClose (node:internal/child_process:1105:16)
    at Socket.<anonymous> (node:internal/child_process:457:11)
    at Socket.emit (node:events:519:28)
    at Pipe.<anonymous> (node:net:338:12) {
  code: 128,
  killed: false,
  signal: null,
  cmd: 'nix-prefetch-git --out csarven_green_turtle.tmp --url https://github.com/csarven/green-turtle --rev 081ae1c39591e7502ac5f9c121253e18122d2a71 --builder',
  stdout: '',
  stderr: "fatal: cannot copy '/nix/store/p45c31p2qhwjy4vf60v2d2mw9bvvpb67-git-2.45.2/share/git-core/templates/hooks/commit-msg.sample' to '/nix/store/9nyg6qj21776n21jvrcww9q0bi22bwxh-offline/csarven_green_turtle.tmp/.git/hooks/commit-msg.sample': File exists\n"
}
```

For reference about my packaging attempt and the issue:
- broken state, using default nixpkgs: `nix build github:feathecutie/dokieli-flake/broken` (https://github.com/feathecutie/dokieli-flake/tree/broken)
- working state, using my nixpkgs fork: `nix build github:feathecutie/dokieli-flake` (https://github.com/feathecutie/dokieli-flake/tree/main)

I'm unsure if this is the right level to fix this at, since this issue seems to occur in every situation where `nix-prefetch-git` is simply run twice, so it might make more sense to fix `nix-prefetch-git` instead of fixing `fetchYarnDeps`.

However, even disregarding the `nix-prefetch-git` issue, this change removes unnecessary duplicate work, so I feel like it would still be reasonable to implement this.

I would appreciate some feedback on this because I have never contributed to such an integral part of nixpkgs before and I'm not sure how to properly test that this change has no unwanted side effects.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by eclairevoyant** [comment]
> This issue seems to occur in every situation where `nix-prefetch-git` is simply run twice

Do you have a reproducer for that? I can't get such an error from simply running the command twice.

**Comment by feathecutie** [comment]
I realized that the specific error I'm getting might be happening because during `fetchYarnDeps` both `nix-prefetch-git` commands are running in parallel, but `nix-prefetch-git` is still failing (i.e. returning a non-0 exit code) for me if I run the following command twice in sequence:
```sh
nix-prefetch-git --out csarven_green_turtle.tmp --url https://github.com/csarven/green-turtle --rev 081ae1c39591e7502ac5f9c121253e18122d2a71 --builder
```

The command is failing with the following error: https://gist.github.com/feathecutie/6572953cb612c121083adaae52849091

It makes sense that `nix-prefetch-git` exits with a non-zero error code here, since the directory it's trying to fetch simply already exists, so this might very well be intended behaviour.
Still, it currently breaks `fetchYarnDeps`, so either of these tools have to be fixed.

**Comment by feathecutie** [comment]
I'm specifically talking about `nix-prefetch-git` from the `nix-prefetch-scripts` package btw since this is the once `fetchYarnDeps` is using, in case that makes a difference.

**Comment by feathecutie** [comment]
This is a minimal repro for the general `fetchYarnDeps` issue I'm trying to fix:

```nix
{ writeText, fetchYarnDeps, ... }:
let
  yarnLock = writeText "yarn.lock" ''
    "green-turtle@git://github.com/bergos/green-turtle.git#master", "green-turtle@https://github.com/csarven/green-turtle#master":
      version "1.4.0"
      resolved "https://github.com/csarven/green-turtle#081ae1c39591e7502ac5f9c121253e18122d2a71"
  '';
in
fetchYarnDeps { inherit yarnLock; }
```

This fails with normal nixpkgs, but suceeds with my fix.

**Review by doronbehar — COMMENTED** [review]
Looks good to me overall.

File: pkgs/build-support/node/fetch-yarn-deps/index.js [hunk 1]
```diff
@@ -143,8 +143,17 @@ const performParallel = tasks => {
...
 	const lockData = lockfile.parse(lockContents)
+	// This could be implemented using [`Map.groupBy`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/groupBy),
+	// but that method is only supported starting with Node 21
+	const uniqueBy = (arr, callback) => {
+		const map = new Map()
+		for (const elem of arr) {
+			map.set(callback(elem), elem)
+		}
+		return [...map.values()]
+	}
```
Comment (doronbehar) [inline pkgs/build-support/node/fetch-yarn-deps/index.js:154]: Could you please define this function outside the `const prefetchYarnDeps` function - it should improve performance.

Comment (feathecutie) [reply pkgs/build-support/node/fetch-yarn-deps/index.js]: Yeah makes sense, done

**Comment by doronbehar** [comment]
Diff looks good. Took a random package to test that nothing is horribly broken due to this:

```
nix build -Lf. vim-language-server.offlineCache --rebuild
```

And it passed. Thanks.

**Pull request merged**